src/ir/instance_fact.h

```diff
@@ -0,0 +1,119 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+#ifndef SRC_IR_INSTANCE_FACT_H_
+#define SRC_IR_INSTANCE_FACT_H_
+
+#include "src/ir/instantiator.h"
+#include "src/ir/noop_instantiator.h"
+
+// An InstanceFact is an instantation of some other fact that may be
```

Comment:
If I understand the PR correctly, we are keeping a `pair<Fact, Instantiator>`, where the `Instantiator` has enough information to create an instance of the `Fact` later as needed. Is that right? 

Given that the classes we have are basically a part of some AST,  all of this seems more complicated than necessary. What do we gain by using the instantiators because we still need to instantiate them later as needed?

Reply:
Not quite. We don't create an instance later, the pair of the `Instantiator` and the original fact *is* the instance. We just use the `Instantiator` to substitute the `AccessPathRoot` on the fly while performing operations where we need the instantiation instead of the spec version.

Previously, we had a spec versions of claims, checks, and edges on `ParticleSpec`s that had one or more `AccessPath`s with `AccessPathRoot`s indicating the `ParticleSpec`. If you accidentally called `ToDatalog()` on that fact, would get a runtime error. For each Particle, we would create an instantiated copy that was almost identical to the original spec fact but with the spec `AccessPathRoot` replaced with an instance `AccessPathRoot`. If we had N particles, we would create N copies, and the only difference between those copies would be the `AccessPathRoot`s in the contained `AccessPath`s.

As of this PR, we create no copies for instances. We augment certain methods that expect to only be called on instantiated facts (such as `ToDatalog`) to take an `Instantiator`, and they run on the original uninstantiated spec fact. The difference is that when these methods encounter an `AccessPathRoot`, they perform whatever operation they intended to perform upon the replacement `AccessPathRoot` provided by the `Instantiator` instead of the original one.

Not making copies greatly simplifies the ownership story if these facts have a `unique_ptr` hanging off of them, which will be the case once `Predicate` trees are added to claims. In addition, it addresses some of the concerns you had about making copies of variable-sized `AccessPath`s everywhere.